File: python/day25.py
from aoc import *


digits = "=-012"

def from_5(number):
    s = 0
    for d in number:
        s *= 5
        s += digits.find(d) - 2
    return s


def to_5(number):
    acc = []
    d = number
    while d:
        d, m = divmod(d+2, 5)
        acc.append(digits[m])
    return cat(reversed(acc))


def solve(filename=25):
    numbers = read_input(filename)
    return to_5(sum(from_5(n) for n in numbers))


print(solve())










def test():
    assert solve("25_test") == "2=-1=0"
    assert solve() == "122-2=200-0111--=200"

File: clojure/day25.clj
(ns day25
  (:require aoc
            [clojure.string :as str]))


(def digits "=-012")


(defn from-5 [number]
  (reduce
   (fn [acc d]
     (+ (* 5 acc)
        (- (str/index-of digits d) 2)))
   0
   number))

(defn to-5 [number]
  (loop [d number
         acc '()]
    (if (zero? d) (str/join acc)
        (let [m (mod  (+ d 2) 5)
              d (quot (+ d 2) 5)
              digit (nth digits m)]
          (recur d (cons digit acc))))))

(defn solve
  ([] (solve (aoc/read-file 25)))
  ([input]
   (let [numbers (aoc/parse-input input)]
     (->> numbers
          (map from-5)
          (reduce +)
          to-5))))


(solve)
